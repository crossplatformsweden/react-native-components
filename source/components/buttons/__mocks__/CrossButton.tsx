export const CrossButton = 'Button';

export default CrossButton;
