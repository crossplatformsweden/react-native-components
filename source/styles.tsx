import { Dimensions, Platform, StyleSheet } from 'react-native';
import { DefaultTheme } from 'react-native-paper';
const { width, height } = Dimensions.get('window');

const screenWidth = width < height ? width : height;

export enum Colors {
  CrossLightBlue = '#057dc2',
  CrossDarkBlue = '#00539b',
  CrossYellow = '#fede00',
  CrossBlack = '#2f2f2f',
  CrossWhite = 'white',
  CancelButton = '#ff0000',
  NextButton = '#349b34',
  BackButton = '#3180e0',
}

/**
 * Size if icons in tab bars
 * @type int
 * @public
 */
export const TabIconSize: number = Platform.OS === 'android' ? 22 : 22;

const IS_IOS = Platform.OS === 'ios';
const { width: viewportWidth, height: viewportHeight } = Dimensions.get(
  'window'
);

function wp(percentage: number) {
  const value = (percentage * viewportWidth) / 100;
  return Math.round(value);
}

export const slideHeight = viewportHeight * 0.15;
const itemWidth = wp(70);
const itemHorizontalMargin = wp(2);

export const sliderWidth = viewportWidth;
export const slideWidth = itemWidth + itemHorizontalMargin * 2;

const entryBorderRadius = 8;

/**
 * Theme appiled to React-Native-Paper
 *
 * https://callstack.github.io/react-native-paper/theming.html
 */
export const Theme = {
  ...DefaultTheme,
  roundness: 4,
  dark: false,
  colors: {
    ...DefaultTheme.colors,
    primary: Colors.CrossLightBlue,
    accent: Colors.CrossYellow,
    background: 'white',
    placeholder: Colors.CrossDarkBlue,
    backdrop: 'white',
    disabled: Colors.CrossDarkBlue,
    text: Colors.CrossBlack,
  },
};

export const CarouselStyles = StyleSheet.create({
  imageContainer: {
    flex: 1,
    marginBottom: IS_IOS ? 0 : -1, // Prevent a random Android rendering issue
    backgroundColor: 'white',
    borderTopLeftRadius: entryBorderRadius,
    borderTopRightRadius: entryBorderRadius,
  },
  image: {
    ...StyleSheet.absoluteFillObject,
    resizeMode: 'cover',
    borderRadius: IS_IOS ? entryBorderRadius : 0,
    borderTopLeftRadius: entryBorderRadius,
    borderTopRightRadius: entryBorderRadius,
  },
  // image's border radius is buggy on iOS; let's hack it!
  radiusMask: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    height: entryBorderRadius,
    backgroundColor: 'white',
  },
});

/**
 * Main theme for the app
 */
export const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
  },
  surface: {
    padding: 8,
    height: 80,
    width: 80,
    alignItems: 'center',
    justifyContent: 'center',
    elevation: 4,
  },
  buttonList: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    alignItems: 'flex-start',
  },
  listButton: {
    height: 40,
    flex: 1,
  },
  listItem: {
    margin: 0,
    padding: 0,
  },
  button: {
    marginRight: 10,
    height: 45,
  },
  paddingDefault: {
    paddingTop: 20,
    paddingLeft: 10,
    paddingRight: 10,
  },
  label: {
    color: 'gray',
    fontSize: 14,
  },
  link: {
    color: 'blue',
    fontWeight: 'bold',
  },
  tabIcon: {
    // Icons in tab bar won't show without explicit width / height
    width: 25,
    height: TabIconSize,
    color: 'white',
  },
  absoluteCentered: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    alignItems: 'center',
    justifyContent: 'center',
  },
  absoluteTop: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  absoluteTopRight: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    alignItems: 'flex-end',
    justifyContent: 'flex-start',
  },
  absoluteTopLeft: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  absoluteBottom: {
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  absoluteBottomRight: {
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
    alignItems: 'flex-end',
    justifyContent: 'flex-end',
  },
  absoluteBottomLeft: {
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
    alignItems: 'flex-start',
    justifyContent: 'flex-end',
  },
  centerVerticalAndHorizontal: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  alignCenterJustifyStart: {
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  alignStartJustifyStart: {
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  alignEndJustifyStart: {
    alignItems: 'flex-end',
    justifyContent: 'flex-start',
  },
  tabBar: {
    backgroundColor: Colors.CrossLightBlue,
  },
  tabBarTitle: {
    alignSelf: 'center',
  },
  tabBarLabel: {
    fontSize: 9,
  },
  rowContentTopCenter: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'flex-start',
  },
  rowContentTopLeft: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  rowContentTopRight: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'flex-start',
  },
  rowContentCenterLeft: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  rowContentBottomCenter: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'flex-end',
  },
  columnContentTopCenter: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'center',
  },
  columnContentTopLeft: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  columnContentTopRight: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-end',
  },
  columnContentCenterRight: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'flex-end',
  },
  columnContentCenterLeft: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'flex-start',
  },
  backgroundWhite: {
    backgroundColor: '#ffffff',
  },
  overlay: {
    backgroundColor: 'rgba(254, 222, 0, 0.3)',
    position: 'absolute',
    left: 0,
    top: 0,
    bottom: 0,
    right: 0,
  },
  spinner: {
    width: 50,
    height: 50,
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'flex-end',
    opacity: 1,
  },
  textCenter: {
    textAlign: 'center',
  },
  textSpinner: {
    fontSize: 20,
    color: Colors.CrossLightBlue,
    textAlign: 'center',
    fontWeight: 'normal',
    marginTop: 5,
    opacity: 1,
  },
  modal: {
    margin: 30,
    flex: 1,
    padding: 0,
    marginTop: 200,
  },
});

export const autoCompleteStyles = StyleSheet.create({
  // container: {
  //   backgroundColor: '#ffffff',
  // },
  padding: { padding: 20 },
  listView: {
    backgroundColor: '#ffffff',
  },
  textInputContainer: {
    borderBottomWidth: 0,
    borderTopWidth: 0,
  },
  textInput: {
    color: '#00539b',
    fontSize: 16,
  },
  description: {
    fontWeight: 'bold',
  },
  predefinedPlacesDescription: {
    color: '#fede00',
  },
});

export const sortableGridImagesStyles = StyleSheet.create({
  block: {
    flex: 1,
    margin: 8.5,
    borderRadius: 10,
    justifyContent: 'center',
    alignItems: 'center',
  },
  contentContainer: {
    flex: 9,
  },
  buttonView: {
    justifyContent: 'space-between',
    flexDirection: 'row',
    marginHorizontal: 15,
  },
  button: {
    width: 150,
    height: 50,
    alignItems: 'center',
    justifyContent: 'center',
  },
  imageButton: {
    flexDirection: 'row',
  },
  button1: {
    backgroundColor: '#257FE6',
  },
  button2: {
    backgroundColor: '#37AB33',
  },
  buttonText: {
    color: '#fff',
  },
  bottom: {
    width: screenWidth,
    alignSelf: 'center',
    flex: 1,
  },
  cam: {
    width: 16,
    height: 16,
    resizeMode: 'contain',
    marginRight: 2,
  },
});

export default styles;
