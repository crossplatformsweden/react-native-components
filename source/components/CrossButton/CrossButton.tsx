import React from 'react';
// @ts-ignore
import { Button } from 'react-native-paper';
import { ViewStyle, StyleProp, View, ButtonProps } from 'react-native';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import { Color } from 'csstype';
import _ from 'lodash';
import styles, { Colors } from '../../styles';

/**
 * Cross button options
 * @typedef ICrossButtonProps
 */
export interface ICrossButtonProps {
  /**
   * Foreground (icon) color. Default is {@link Colors.CrossBlack}
   */
  color?: Color;
  /**
   * Font-Awesome icon name
   *
   * https://fontawesome.com/icons/
   */
  iconName?: string;
  /**
   * Background color. Default value is {@link Colors.NextButton} if {@link title} was supplied
   */
  backgroundColor?: Color;
  style?: StyleProp<ViewStyle>;
  title?: string;
  /**
   * Size of the icon
   */
  size?: number;
  disabled?: boolean;
  /**
   * Button visual appearence. Default is 'text'
   *
   * https://callstack.github.io/react-native-paper/button.html
   */
  mode?: 'text' | 'outlined' | 'contained';
  /**
   * Optional button styles to apply
   */
  buttonStyle?: ButtonProps;
  onPress?: () => void;
  compact?: boolean;
}

/**
 * A custom button that displays as an icon if {@param title} is not supplied
 *
 * Remarks: use custom {@param onPressButton} event
 * @param param0 {@link ICrossButtonProps}
 */
export class CrossButton extends React.Component<ICrossButtonProps> {
  render() {
    const mode = this.props.mode || 'text';
    return (
      <View style={styles.container}>
        {_.isNil(this.props.title) && !_.isNil(this.props.iconName) ? (
          <FontAwesome.Button
            {...this.props}
            style={(this.props.style as ViewStyle) || styles.container}
            onPress={() => {
              if (this.props.onPress) {
                this.props.onPress();
              }
            }}
            color={this.props.color || Colors.CrossBlack}
            name={this.props.iconName.toString()
            }
          />
        ) : (
          <Button
            color={this.props.backgroundColor || Colors.NextButton}
            disabled={this.props.disabled || false}
            {...this.props}
            icon={({ size, color }) =>
              this.props.iconName ? (
                <FontAwesome
                  name={this.props.iconName}
                  size={this.props.size || size}
                  style={{ margin: 0, padding: 0 }}
                  color={color}
                />
              ) : null
            }
            style={this.props.style || this.props.buttonStyle || styles.button}
            onPress={
              this.props.onPress
                ? this.props.onPress
                : () => console.log('** CrossButton press **')
            }
            mode={mode}
          >
            {this.props.title}
          </Button>
        )}
      </View>
    );
  }
}

export default CrossButton;
