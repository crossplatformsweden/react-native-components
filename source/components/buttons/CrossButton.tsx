import React from 'react';
// @ts-ignore
import { Button } from 'react-native-paper';
import { ViewStyle, StyleProp, View, TextStyle } from 'react-native';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import { Color } from 'csstype';
import _ from 'lodash';
import styles, { Colors } from '../../styles';

/**
 * Properties for the {@link CrossButton} component
 * @typedef ICrossButtonProps
 * @type ICrossButtonProps
 */
export interface ICrossButtonProps {
  /**
   * Foreground (icon) color. Default is {@link Colors.CrossBlack}
   */
  color?: Color;
  /**
   * Font-Awesome 4 icon name using `react-native-vector-icons/FontAwesome`
   *
   * https://fontawesome.com/v4.7.0/icons/
   */
  iconName?: string;
  /**
   * Background color. Default value is {@link Colors.NextButton} if {@link title} was supplied
   */
  backgroundColor?: Color;
  /**
   * Style for outer container.
   *
   * See also {@link iconStyle}, {@link buttonStyle}
   */
  style?: StyleProp<ViewStyle>;
  /**
   * Optional style for the icons. See also {@link iconName}
   *
   * See also {@link buttonStyle}, {@link style}
   */
  iconStyle?: ViewStyle | TextStyle;
  /**
   * Title for the button. If not supplied a clickable icon will be displayed instead of react-native-paper button
   *
   * https://callstack.github.io/react-native-paper/button.html
   */
  title?: string;
  /**
   * Size of the icon
   */
  size?: number;
  disabled?: boolean;
  /**
   * Button visual appearence. Default is 'text'
   *
   * https://callstack.github.io/react-native-paper/button.html
   */
  mode?: 'text' | 'outlined' | 'contained';
  /**
   * Optional button styles to apply to {@link https://callstack.github.io/react-native-paper/button.html Button}.
   *
   * See also {@link iconStyle}, {@link style}
   */
  buttonStyle?: StyleProp<ViewStyle>;
  onPress?: () => void;
  compact?: boolean;
}

/**
 * A custom button that displays as an Paper Button with icon or just an icon if {@link ICrossButtonProps.title} is not supplied.
 *
 * Remarks: all icons are {@link https://fontawesome.com/v4.7.0/icons/ FontAwesome v4}.
 *
 * Remark: default appearence is "text", see {@link ICrossButtonProps.mode}
 *
 * Properties are {@link ICrossButtonProps}
 */
export class CrossButton extends React.Component<ICrossButtonProps> {
  render() {
    const mode = this.props.mode || 'text';
    return (
      <View style={this.props.style ? this.props.style : styles.container}>
        {_.isNil(this.props.title) && !_.isNil(this.props.iconName) ? (
          <FontAwesome.Button
            {...this.props}
            style={this.props.iconStyle || styles.container as ViewStyle}
            onPress={() => {
              if (this.props.onPress) {
                this.props.onPress();
              }
            }}
            color={this.props.color || Colors.CrossBlack}
            name={this.props.iconName.toString()}
          />
        ) : (
          <Button
            color={this.props.backgroundColor || Colors.NextButton}
            disabled={this.props.disabled || false}
            {...this.props}
            icon={({ size, color }) =>
              this.props.iconName ? (
                <FontAwesome
                  name={this.props.iconName}
                  size={this.props.size || size}
                  style={{ margin: 0, padding: 0 }}
                  color={color}
                />
              ) : null
            }
            style={this.props.buttonStyle || styles.button}
            onPress={
              this.props.onPress
                ? this.props.onPress
                : () => console.log('** CrossButton press **')
            }
            mode={mode}
          >
            {this.props.title}
          </Button>
        )}
      </View>
    );
  }
}

export default CrossButton;
