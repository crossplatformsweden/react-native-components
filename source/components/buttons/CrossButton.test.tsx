import React from 'react';
import CrossButton, { ICrossButtonProps } from './CrossButton';
import TestRenderer from 'react-test-renderer';
import { styles } from '../../styles';

function setup(
  iconName: string,
  title?: any,
  size?: number,
  onPress?: () => void
) {
  return TestRenderer.create(
    <CrossButton
      iconName={iconName}
      title={title}
      color="blue"
      onPress={onPress || jest.fn()}
      size={size || 64}
    />
  );
}

jest.unmock('react-native');
jest.unmock('./CrossButton');

describe('components', () => {
  describe('<CrossButton />', () => {
    describe('`ICrossButtonProps` interface', () => {
      const propsMock: ICrossButtonProps = {};
      it('Should match snapshot', () => {
        expect(propsMock).toMatchSnapshot();
      });

      it('Should have `color` property that is undefined', () => {
        expect(propsMock.color).toBeUndefined();
      });
    });

    describe('With `title` and `iconName` properties provided', () => {
      it('Component should render', () => {
        const wrapper = setup('home', 'home');
        expect(wrapper.toJSON()).toMatchSnapshot();
      });

      it('Should have paper <Button />', () => {
        const wrapper = setup('home', 'home');
        const child = wrapper.root.findByProps({ mode: 'text' });
        expect(child.type).toEqual('Button');
      });

      it('Should have paper `<Button />` with `mode` property set to "text"', () => {
        const wrapper = setup('home', 'home');
        const child = wrapper.root.findByProps({ mode: 'text' });
        expect(child).toBeDefined();
      });

      it('Should have paper `<FontAwesome />` icon with `size` property set to `99`', () => {
        const wrapper = setup('home', 'home', 99);
        const child = wrapper.root.findByProps({ size: 99 });
        expect(child.props.color).toBe('blue');
      });

      it('Should have paper `<FontAwesome />` icon with `size` property set to `99`', () => {
        const wrapper = setup('home', 'home', 99);
        const child = wrapper.root.findByProps({ size: 99 });
        expect(child).toBeDefined();
      });
    });

    describe('With `iconName` but without `title` property provided', () => {
      it('Component should render', () => {
        const wrapper = setup('map', undefined);
        expect(wrapper.toJSON()).toMatchSnapshot();
      });

      it('Should have `<FontAwesome.Button />` with `name` set to "map"', () => {
        const wrapper = setup('map', undefined);
        const child = wrapper.root.findByProps({ name: 'map' });
        expect(child).toBeDefined();
      });

      it('Should have `<FontAwesome.Button />` with `style` set to "container"', () => {
        const wrapper = setup('map', undefined);
        const child = wrapper.root.findByProps({ name: 'map' });
        expect(child.props.style).toBe(styles.container);
      });

      it('<FontAwesome.Button /> onPress should be called', () => {
        let called = false;
        const onCalled = () => (called = !called);

        const wrapper = setup('map', undefined, undefined, onCalled);
        const child = wrapper.root.findByProps({
          testID: 'FontAwesomeButton',
        });
        child.props.onPress();
        expect(called).toBeTruthy();
      });
    });
  });
});
