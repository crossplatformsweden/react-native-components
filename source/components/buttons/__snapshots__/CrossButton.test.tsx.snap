// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <CrossButton /> \`ICrossButtonProps\` interface Should match snapshot 1`] = `Object {}`;

exports[`components <CrossButton /> With \`iconName\` but without \`title\` property provided Component should render 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    color="blue"
    iconName="map"
    name="map"
    onPress={[Function]}
    size={64}
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
      }
    }
  />
</View>
`;

exports[`components <CrossButton /> With \`title\` and \`iconName\` properties provided Component should render 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <Button
    color="blue"
    disabled={false}
    icon={[Function]}
    iconName="home"
    mode="text"
    onPress={[MockFunction]}
    size={64}
    style={
      Object {
        "height": 45,
        "marginRight": 10,
      }
    }
    title="home"
  >
    home
  </Button>
</View>
`;
